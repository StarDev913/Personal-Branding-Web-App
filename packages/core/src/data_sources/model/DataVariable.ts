import { Model } from '../../common';
import EditorModel from '../../editor/model/Editor';
import { isDataVariable } from '../utils';
import { DataCollectionStateMap, DataCollectionState, DataCollectionStateType } from './data_collection/types';

export const DataVariableType = 'data-variable' as const;

export interface DataVariableProps {
  type?: typeof DataVariableType;
  path?: string;
  defaultValue?: string;
  collectionId?: string;
  variableType?: DataCollectionStateType;
  asPlainText?: boolean;
}

interface DataVariableOptions {
  em: EditorModel;
  collectionsStateMap: DataCollectionStateMap;
}

export default class DataVariable extends Model<DataVariableProps> {
  private em: EditorModel;
  private collectionsStateMap: DataCollectionStateMap;

  defaults(): DataVariableProps {
    return {
      type: DataVariableType,
      defaultValue: '',
      path: '',
      collectionId: undefined,
      variableType: undefined,
      asPlainText: undefined,
    };
  }

  constructor(props: DataVariableProps, options: DataVariableOptions) {
    super(props, options);
    this.em = options.em;
    this.collectionsStateMap = options.collectionsStateMap;
  }

  get path(): string {
    return this.get('path') ?? '';
  }

  get defaultValue(): string {
    return this.get('defaultValue') ?? '';
  }

  get collectionId(): string | undefined {
    return this.get('collectionId');
  }

  get variableType(): DataCollectionStateType | undefined {
    return this.get('variableType');
  }

  resolvesFromCollection(): boolean {
    return !!this.collectionId;
  }

  updateCollectionsStateMap(collectionsStateMap: DataCollectionStateMap): void {
    this.collectionsStateMap = collectionsStateMap;
    this.trigger('change');
  }

  getResolverPath(): string | false {
    if (this.resolvesFromCollection()) {
      const value = this.resolveCollectionVariable();
      return isDataVariable(value) ? (value.path ?? '') : false;
    }
    return this.path;
  }

  toJSON(options?: any): DataVariableProps & { type: typeof DataVariableType } {
    const defaults = this.defaults();
    const json = super.toJSON(options);
    const filteredJson = Object.fromEntries(
      Object.entries(json).filter(([key, value]) => value !== defaults[key as keyof DataVariableProps]),
    ) as Partial<DataVariableProps>;
    return { ...filteredJson, type: DataVariableType };
  }

  getDataValue() {
    const opts = {
      em: this.em,
      collectionsStateMap: this.collectionsStateMap,
    };

    return DataVariable.resolveDataResolver(
      {
        path: this.path,
        defaultValue: this.defaultValue,
        collectionId: this.collectionId,
        variableType: this.variableType,
      },
      opts,
    );
  }

  static resolveDataSourceVariable(
    props: {
      path?: string;
      defaultValue?: string;
    },
    opts: {
      em: EditorModel;
    },
  ) {
    return opts.em.DataSources.getValue(props.path ?? '', props.defaultValue ?? '');
  }

  static resolveDataResolver(
    props: {
      path?: string;
      defaultValue?: string;
      collectionId?: string;
      variableType?: DataCollectionStateType;
    },
    opts: DataVariableOptions,
  ) {
    if (props.collectionId) {
      const value = DataVariable.resolveCollectionVariable(props, opts);
      if (!isDataVariable(value)) return value;
      return DataVariable.resolveDataSourceVariable(
        { path: value.path ?? '', defaultValue: props.defaultValue ?? '' },
        { em: opts.em },
      );
    }
    return DataVariable.resolveDataSourceVariable(
      { path: props.path ?? '', defaultValue: props.defaultValue ?? '' },
      { em: opts.em },
    );
  }

  private resolveCollectionVariable(): unknown {
    const { em, collectionsStateMap } = this;
    return DataVariable.resolveCollectionVariable(this.attributes, { em, collectionsStateMap });
  }

  static resolveCollectionVariable(
    dataResolverProps: {
      collectionId?: string;
      variableType?: DataCollectionStateType;
      path?: string;
      defaultValue?: string;
    },
    opts: DataVariableOptions,
  ): unknown {
    const { collectionId = '', variableType, path, defaultValue = '' } = dataResolverProps;
    const { em, collectionsStateMap } = opts;

    if (!collectionsStateMap) return defaultValue;

    const collectionItem = collectionsStateMap[collectionId];
    if (!collectionItem) return defaultValue;

    if (!variableType) {
      em.logError(`Missing collection variable type for collection: ${collectionId}`);
      return defaultValue;
    }

    return variableType === 'currentItem'
      ? DataVariable.resolveCurrentItem(collectionItem, path, collectionId, em)
      : collectionItem[variableType];
  }

  private static resolveCurrentItem(
    collectionItem: DataCollectionState,
    path: string | undefined,
    collectionId: string,
    em: EditorModel,
  ): unknown {
    const currentItem = collectionItem.currentItem;
    if (!currentItem) {
      em.logError(`Current item is missing for collection: ${collectionId}`);
      return '';
    }

    if (currentItem.type === DataVariableType) {
      const resolvedPath = currentItem.path ? `${currentItem.path}.${path}` : path;
      return { type: DataVariableType, path: resolvedPath };
    }

    if (path && !(currentItem as any)[path]) {
      em.logError(`Path not found in current item: ${path} for collection: ${collectionId}`);
      return '';
    }

    return path ? (currentItem as any)[path] : currentItem;
  }
}
