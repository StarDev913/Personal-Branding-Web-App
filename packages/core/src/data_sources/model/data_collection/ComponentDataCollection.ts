import { isArray } from 'underscore';
import { ObjectAny } from '../../../common';
import Component, { keySymbol } from '../../../dom_components/model/Component';
import { ComponentAddType, ComponentDefinitionDefined, ComponentOptions } from '../../../dom_components/model/types';
import EditorModel from '../../../editor/model/Editor';
import { isObject, toLowerCase } from '../../../utils/mixins';
import DataResolverListener from '../DataResolverListener';
import DataSource from '../DataSource';
import DataVariable, { DataVariableProps, DataVariableType } from '../DataVariable';
import { isDataVariable } from '../../utils';
import { DataCollectionItemType, DataCollectionType, keyCollectionDefinition } from './constants';
import {
  ComponentDataCollectionProps,
  DataCollectionDataSource,
  DataCollectionProps,
  DataCollectionStateMap,
} from './types';
import { detachSymbolInstance, getSymbolInstances } from '../../../dom_components/model/SymbolUtils';
import { updateFromWatcher } from '../../../dom_components/model/ComponentDataResolverWatchers';
import { ModelDestroyOptions } from 'backbone';
import Components from '../../../dom_components/model/Components';

const AvoidStoreOptions = { avoidStore: true, partial: true };
type DataVariableMap = Record<string, DataVariableProps>;

export default class ComponentDataCollection extends Component {
  dataSourceWatcher?: DataResolverListener;

  get defaults(): ComponentDefinitionDefined {
    return {
      // @ts-ignore
      ...super.defaults,
      droppable: false,
      dataResolver: {},
      type: DataCollectionType,
      components: [
        {
          type: DataCollectionItemType,
        },
      ],
    };
  }

  constructor(props: ComponentDataCollectionProps, opt: ComponentOptions) {
    if (opt.forCloning) {
      return super(props as any, opt) as unknown as ComponentDataCollection;
    }

    const newProps = { ...props, droppable: false } as any;
    const cmp: ComponentDataCollection = super(newProps, opt) as unknown as ComponentDataCollection;
    this.rebuildChildrenFromCollection = this.rebuildChildrenFromCollection.bind(this);
    this.listenToPropsChange();
    this.rebuildChildrenFromCollection();

    return cmp;
  }

  getDataResolver() {
    return this.get('dataResolver');
  }

  getItemsCount() {
    const items = this.getDataSourceItems();
    const itemsCount = getLength(items);

    const startIndex = Math.max(0, this.getConfigStartIndex() ?? 0);
    const configEndIndex = this.getConfigEndIndex() ?? Number.MAX_VALUE;
    const endIndex = Math.min(itemsCount - 1, configEndIndex);

    const count = endIndex - startIndex + 1;
    return Math.max(0, count);
  }

  getConfigStartIndex() {
    return this.dataResolver.startIndex;
  }

  getConfigEndIndex() {
    return this.dataResolver.endIndex;
  }

  getDataSource(): DataCollectionDataSource {
    return this.dataResolver?.dataSource;
  }

  getCollectionId(): string {
    return this.dataResolver?.collectionId;
  }

  getCollectionItemComponents() {
    return this.firstChild.components();
  }

  setDataResolver(props: DataCollectionProps) {
    return this.set('dataResolver', props);
  }

  setCollectionId(collectionId: string) {
    this.updateCollectionConfig({ collectionId });
  }

  setStartIndex(startIndex: number): void {
    if (startIndex < 0) {
      this.em.logError('Start index should be greater than or equal to 0');
      return;
    }

    this.updateCollectionConfig({ startIndex });
  }

  setEndIndex(endIndex: number): void {
    this.updateCollectionConfig({ endIndex });
  }

  setDataSource(dataSource: DataCollectionDataSource) {
    this.set(keyCollectionDefinition, {
      ...this.dataResolver,
      dataSource,
    });
  }

  setCollectionItemComponents(content: ComponentAddType) {
    this.firstChild.components(content);
  }

  private get firstChild() {
    return this.components().at(0);
  }

  private updateCollectionConfig(updates: Partial<DataCollectionProps>): void {
    this.set(keyCollectionDefinition, {
      ...this.dataResolver,
      ...updates,
    });
  }

  private getDataSourceItems() {
    const items = getDataSourceItems(this.dataResolver.dataSource, this.em);
    if (isArray(items)) {
      return items;
    }

    const clone = { ...items };
    delete clone['__p'];
    return clone;
  }

  private get dataResolver() {
    return (this.get(keyCollectionDefinition) || {}) as DataCollectionProps;
  }

  private get collectionDataSource() {
    return this.dataResolver.dataSource;
  }

  private listenToDataSource() {
    const { em } = this;
    const path = this.collectionDataSource?.path;
    if (!path) return;
    this.dataSourceWatcher = new DataResolverListener({
      em,
      resolver: new DataVariable(
        { type: DataVariableType, path },
        { em, collectionsStateMap: this.collectionsStateMap },
      ),
      onUpdate: this.rebuildChildrenFromCollection,
    });
  }

  private rebuildChildrenFromCollection() {
    this.components().reset(this.getCollectionItems(), updateFromWatcher as any);
  }

  private getCollectionItems() {
    const firstChild = this.ensureFirstChild();
    const displayStyle = firstChild.getStyle()['display'];
    const isDisplayNoneOrMissing = !displayStyle || displayStyle === 'none';
    const resolvedDisplay = isDisplayNoneOrMissing ? '' : displayStyle;
    // TODO: Move to component view
    firstChild.addStyle({ display: 'none' }, AvoidStoreOptions);
    const components: Component[] = [firstChild];

    const result = validateCollectionDef(this.dataResolver, this.em);
    if (!result) {
      return components;
    }

    const collectionId = this.collectionId;
    const items = this.getDataSourceItems();
    const { startIndex, endIndex } = this.resolveCollectionConfig(items);

    const isDuplicatedId = this.hasDuplicateCollectionId();
    if (isDuplicatedId) {
      this.em.logError(
        `The collection ID "${collectionId}" already exists in the parent collection state. Overriding it is not allowed.`,
      );

      return components;
    }

    for (let index = startIndex; index <= endIndex; index++) {
      const isFirstItem = index === startIndex;
      const key = isArray(items) ? index : Object.keys(items)[index];
      const collectionsStateMap = this.getCollectionsStateMapForItem(items, key);

      if (isFirstItem) {
        getSymbolInstances(firstChild)?.forEach((cmp) => detachSymbolInstance(cmp));

        setCollectionStateMapAndPropagate(firstChild, collectionsStateMap);
        // TODO: Move to component view
        firstChild.addStyle({ display: resolvedDisplay }, AvoidStoreOptions);

        continue;
      }

      const instance = firstChild!.clone({ symbol: true, symbolInv: true });
      instance.set({ locked: true, layerable: false }, AvoidStoreOptions);
      setCollectionStateMapAndPropagate(instance, collectionsStateMap);
      components.push(instance);
    }

    return components;
  }

  private getCollectionsStateMapForItem(items: DataVariableProps[] | DataVariableMap, key: number | string) {
    const { startIndex, endIndex, totalItems } = this.resolveCollectionConfig(items);
    const collectionId = this.collectionId;
    let item: DataVariableProps = (items as any)[key];
    const parentCollectionStateMap = this.collectionsStateMap;

    const numericKey = typeof key === 'string' ? Object.keys(items).indexOf(key) : key;
    const offset = numericKey - startIndex;
    const remainingItems = totalItems - (1 + offset);
    const collectionState = {
      collectionId,
      currentIndex: numericKey,
      currentItem: item,
      currentKey: key,
      startIndex,
      endIndex,
      totalItems,
      remainingItems,
    };

    const collectionsStateMap: DataCollectionStateMap = {
      ...parentCollectionStateMap,
      [collectionId]: collectionState,
    };

    return collectionsStateMap;
  }

  private hasDuplicateCollectionId() {
    const collectionId = this.collectionId;
    const parentCollectionStateMap = this.collectionsStateMap;

    return !!parentCollectionStateMap[collectionId];
  }

  private resolveCollectionConfig(items: DataVariableProps[] | DataVariableMap) {
    const isArray = Array.isArray(items);
    const actualItemCount = isArray ? items.length : Object.keys(items).length;

    const startIndex = this.getConfigStartIndex() ?? 0;
    const configEndIndex = this.getConfigEndIndex() ?? Number.MAX_VALUE;
    const endIndex = Math.min(actualItemCount - 1, configEndIndex);

    let totalItems = 0;
    if (actualItemCount > 0) {
      totalItems = Math.max(0, endIndex - startIndex + 1);
    }

    return { startIndex, endIndex, totalItems, isArray };
  }

  private ensureFirstChild() {
    const dataConditionItemModel = this.em.Components.getType(DataCollectionItemType)!.model;

    return (
      this.firstChild ||
      new dataConditionItemModel(
        {
          type: DataCollectionItemType,
        },
        this.opt,
      )
    );
  }

  private listenToPropsChange() {
    this.on(`change:${keyCollectionDefinition}`, () => {
      this.rebuildChildrenFromCollection();
      this.listenToDataSource();
    });
    this.listenToDataSource();
  }

  private removePropsListeners() {
    this.off(`change:${keyCollectionDefinition}`);
    this.dataSourceWatcher?.destroy();
  }

  onCollectionsStateMapUpdate(collectionsStateMap: DataCollectionStateMap) {
    this.collectionsStateMap = collectionsStateMap;
    this.dataResolverWatchers.onCollectionsStateMapUpdate();

    const items = this.getDataSourceItems();
    const { startIndex } = this.resolveCollectionConfig(items);
    const cmps = this.components();
    cmps.forEach((cmp, index) => {
      const collectionsStateMap = this.getCollectionsStateMapForItem(items, startIndex + index);
      cmp.onCollectionsStateMapUpdate(collectionsStateMap);
    });
  }

  stopSyncComponentCollectionState() {
    this.stopListening(this.components(), 'add remove reset', this.syncOnComponentChange);
    this.onCollectionsStateMapUpdate({});
  }

  syncOnComponentChange(model: Component, collection: Components, opts: any) {
    const collectionsStateMap = this.collectionsStateMap;
    // Avoid assigning wrong collectionsStateMap value to children components
    this.collectionsStateMap = {};

    super.syncOnComponentChange(model, collection, opts);
    this.collectionsStateMap = collectionsStateMap;
    this.onCollectionsStateMapUpdate(collectionsStateMap);
  }

  private get collectionId() {
    return this.getDataResolver().collectionId as string;
  }

  static isComponent(el: HTMLElement) {
    return toLowerCase(el.tagName) === DataCollectionType;
  }

  toJSON(opts?: ObjectAny) {
    const json = super.toJSON.call(this, opts) as ComponentDataCollectionProps;
    delete json.droppable;
    delete json[keySymbol];
    delete json.attributes?.id;

    const firstChild = this.firstChild as any;
    return { ...json, components: [firstChild] };
  }

  destroy(options?: ModelDestroyOptions | undefined): false | JQueryXHR {
    this.removePropsListeners();
    return super.destroy(options);
  }
}

function getLength(items: DataVariableProps[] | object) {
  return isArray(items) ? items.length : Object.keys(items).length;
}

function setCollectionStateMapAndPropagate(cmp: Component, collectionsStateMap: DataCollectionStateMap) {
  cmp.setSymbolOverride(['locked', 'layerable']);
  cmp.syncComponentsCollectionState();
  cmp.onCollectionsStateMapUpdate(collectionsStateMap);
}

function logErrorIfMissing(property: any, propertyPath: string, em: EditorModel) {
  if (!property) {
    em.logError(`The "${propertyPath}" property is required in the collection definition.`);
    return false;
  }
  return true;
}

function validateCollectionDef(dataResolver: DataCollectionProps, em: EditorModel) {
  const validations = [
    { property: dataResolver?.collectionId, propertyPath: 'dataResolver.collectionId' },
    { property: dataResolver?.dataSource, propertyPath: 'dataResolver.dataSource' },
  ];

  for (const { propertyPath } of validations) {
    if (!logErrorIfMissing(dataResolver, propertyPath, em)) {
      return [];
    }
  }

  const startIndex = dataResolver?.startIndex;

  if (startIndex !== undefined && (startIndex < 0 || !Number.isInteger(startIndex))) {
    em.logError(`Invalid startIndex: ${startIndex}. It must be a non-negative integer.`);
  }

  return true;
}

function getDataSourceItems(
  dataSource: DataCollectionDataSource,
  em: EditorModel,
): DataVariableProps[] | DataVariableMap {
  switch (true) {
    case isObject(dataSource) && dataSource instanceof DataSource: {
      const id = dataSource.get('id')!;
      return listDataSourceVariables(id, em);
    }
    case isDataVariable(dataSource): {
      const path = dataSource.path;
      if (!path) return [];
      const isDataSourceId = path.split('.').length === 1;
      if (isDataSourceId) {
        return listDataSourceVariables(path, em);
      } else {
        return em.DataSources.getValue(path, []);
      }
    }
    default:
      return [];
  }
}

function listDataSourceVariables(dataSource_id: string, em: EditorModel): DataVariableProps[] {
  const records = em.DataSources.getValue(dataSource_id, []);
  const keys = Object.keys(records);

  return keys.map((key) => ({
    type: DataVariableType,
    path: dataSource_id + '.' + key,
  }));
}
