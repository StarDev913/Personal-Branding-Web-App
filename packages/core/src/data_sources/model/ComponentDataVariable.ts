import { ComponentOptions, ComponentProperties } from '../../dom_components/model/types';
import { toLowerCase } from '../../utils/mixins';
import DataVariable, { DataVariableProps, DataVariableType } from './DataVariable';
import { ComponentWithDataResolver } from './ComponentWithDataResolver';
import { DataResolver } from '../types';
import { DataCollectionStateMap } from './data_collection/types';

export interface ComponentDataVariableProps extends ComponentProperties {
  type?: typeof DataVariableType;
  dataResolver?: DataVariableProps;
}

export default class ComponentDataVariable extends ComponentWithDataResolver<DataVariableProps> {
  get defaults() {
    return {
      // @ts-ignore
      ...super.defaults,
      type: DataVariableType,
      dataResolver: {},
      droppable: false,
    };
  }

  getPath() {
    return this.dataResolver.get('path');
  }

  getCollectionId() {
    return this.dataResolver.get('collectionId');
  }

  getVariableType() {
    return this.dataResolver.get('variableType');
  }

  getDefaultValue() {
    return this.dataResolver.get('defaultValue');
  }

  getDataValue() {
    return this.dataResolver.getDataValue();
  }

  resolvesFromCollection() {
    return this.dataResolver.resolvesFromCollection();
  }

  getInnerHTML() {
    return this.getDataValue();
  }

  setPath(newPath: string) {
    this.dataResolver.set('path', newPath);
  }

  setDefaultValue(newValue: string) {
    this.dataResolver.set('defaultValue', newValue);
  }

  /**
   * Sets the data source path and resets related properties.
   * This will set collectionId and variableType to undefined as it's typically
   * used when changing to a completely different data source.
   * @param newPath The new path to set as the data source
   */
  resetDataSourcePath(newPath: string) {
    this.set('dataResolver', {
      path: newPath,
      collectionId: undefined,
      variableType: undefined,
    });
  }

  protected createResolverInstance(
    props: DataVariableProps,
    options: ComponentOptions & { collectionsStateMap: DataCollectionStateMap },
  ): DataResolver {
    return new DataVariable(props, options);
  }

  static isComponent(el: HTMLElement) {
    return toLowerCase(el.tagName) === DataVariableType;
  }
}
