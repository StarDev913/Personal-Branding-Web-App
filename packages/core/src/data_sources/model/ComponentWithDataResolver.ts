import { ModelDestroyOptions } from 'backbone';
import { ObjectAny } from '../../common';
import Component from '../../dom_components/model/Component';
import { DataResolver, DataResolverProps, ResolverFromProps } from '../types';
import { ComponentOptions, ComponentProperties } from '../../dom_components/model/types';
import { DataCollectionStateMap } from './data_collection/types';

interface ComponentWithDataResolverProps<T extends DataResolverProps> extends ComponentProperties {
  type: T['type'];
  dataResolver: T;
}

export abstract class ComponentWithDataResolver<T extends DataResolverProps> extends Component {
  dataResolver: ResolverFromProps<T>;

  constructor(props: ComponentWithDataResolverProps<T>, opt: ComponentOptions) {
    super(props, opt);

    this.dataResolver = this.initializeDataResolver(props, opt);
    this.listenToPropsChange();
  }

  private initializeDataResolver(
    props: ComponentWithDataResolverProps<T>,
    opt: ComponentOptions,
  ): ResolverFromProps<T> {
    const resolverProps = props.dataResolver ?? {
      type: props.type,
    };

    const resolver = this.createResolverInstance(resolverProps, {
      ...opt,
      collectionsStateMap: this.collectionsStateMap,
    });

    return resolver as ResolverFromProps<T>;
  }

  protected abstract createResolverInstance(
    props: T,
    options: ComponentOptions & { collectionsStateMap: DataCollectionStateMap },
  ): DataResolver;

  getDataResolver(): T {
    return this.get('dataResolver');
  }

  setDataResolver(props: T) {
    return this.set('dataResolver', props);
  }

  onCollectionsStateMapUpdate(collectionsStateMap: DataCollectionStateMap): void {
    this.dataResolver.updateCollectionsStateMap(collectionsStateMap);
    super.onCollectionsStateMapUpdate(collectionsStateMap);
  }

  protected listenToPropsChange() {
    this.listenTo(
      this.dataResolver,
      'change',
      (() => {
        this.__changesUp({ m: this });
      }).bind(this),
    );

    this.on('change:dataResolver', () => {
      // @ts-ignore
      this.dataResolver.set(this.get('dataResolver'));
    });
  }

  protected removePropsListeners() {
    this.stopListening(this.dataResolver);
    this.off('change:dataResolver');
    this.off(`change:${this.collectionsStateMap}`);
  }

  destroy(options?: ModelDestroyOptions | undefined): false | JQueryXHR {
    this.removePropsListeners();
    return super.destroy(options);
  }

  toJSON(opts?: ObjectAny): any {
    const json = super.toJSON(opts);
    const dataResolver = this.dataResolver.toJSON();
    delete dataResolver.type;

    return {
      ...json,
      dataResolver,
    };
  }
}
