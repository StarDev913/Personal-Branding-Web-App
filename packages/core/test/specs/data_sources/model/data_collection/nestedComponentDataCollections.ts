import { Component, DataRecord, DataSource, DataSourceManager, Editor } from '../../../../../src';
import { DataVariableType } from '../../../../../src/data_sources/model/DataVariable';
import ComponentDataCollection from '../../../../../src/data_sources/model/data_collection/ComponentDataCollection';
import {
  DataCollectionItemType,
  DataCollectionType,
} from '../../../../../src/data_sources/model/data_collection/constants';
import {
  ComponentDataCollectionProps,
  DataCollectionStateType,
} from '../../../../../src/data_sources/model/data_collection/types';
import EditorModel from '../../../../../src/editor/model/Editor';
import { setupTestEditor } from '../../../../common';

describe('Collection component', () => {
  let em: EditorModel;
  let dsm: DataSourceManager;
  let dataSource: DataSource;
  let nestedDataSource: DataSource;
  let wrapper: Component;
  let firstRecord: DataRecord;
  let firstNestedRecord: DataRecord;
  let cmpDef: ComponentDataCollectionProps | undefined;
  let nestedCmpDef: ComponentDataCollectionProps | undefined;
  let parentCmp: ComponentDataCollection;
  let nestedCmp: ComponentDataCollection;

  function getCmpDef(nestedCmpDef: ComponentDataCollectionProps): ComponentDataCollectionProps {
    return {
      type: DataCollectionType,
      components: { type: DataCollectionItemType, components: nestedCmpDef },
      dataResolver: {
        collectionId: 'parent_collection',
        dataSource: {
          type: DataVariableType,
          path: 'my_data_source_id',
        },
      },
    };
  }

  beforeEach(() => {
    ({ em, dsm } = setupTestEditor());
    wrapper = em.getWrapper()!;
    dataSource = dsm.add({
      id: 'my_data_source_id',
      records: [
        { id: 'user1', user: 'user1', age: '12' },
        { id: 'user2', user: 'user2', age: '14' },
      ],
    });

    nestedDataSource = dsm.add({
      id: 'nested_data_source_id',
      records: [
        { id: 'nested_user1', user: 'nested_user1', age: '12' },
        { id: 'nested_user2', user: 'nested_user2', age: '14' },
        { id: 'nested_user3', user: 'nested_user3', age: '16' },
      ],
    });

    firstRecord = dataSource.getRecord('user1')!;
    firstNestedRecord = nestedDataSource.getRecord('nested_user1')!;

    nestedCmpDef = {
      type: DataCollectionType,
      components: {
        type: DataCollectionItemType,
        components: {
          type: 'default',
          name: {
            type: DataVariableType,
            variableType: DataCollectionStateType.currentItem,
            collectionId: 'nested_collection',
            path: 'user',
          },
        },
      },
      dataResolver: {
        collectionId: 'nested_collection',
        dataSource: {
          type: DataVariableType,
          path: 'nested_data_source_id',
        },
      },
    };

    cmpDef = getCmpDef(nestedCmpDef);

    parentCmp = wrapper.components(cmpDef)[0] as unknown as ComponentDataCollection;
    nestedCmp = parentCmp.getCollectionItemComponents().at(0) as ComponentDataCollection;
  });

  afterEach(() => {
    em.destroy();
    nestedCmpDef = undefined;
    cmpDef = undefined;
  });

  test('Nested collections bind to correct data sources', () => {
    const nestedFirstChild = nestedCmp.components().at(0).components().at(0);
    const nestedSecondChild = nestedCmp.components().at(1).components().at(0);

    expect(nestedFirstChild.get('name')).toBe('nested_user1');
    expect(nestedSecondChild.get('name')).toBe('nested_user2');
  });

  test('Updates in parent collection propagate to nested collections', () => {
    const nestedFirstChild = nestedCmp.components().at(0).components().at(0);
    const nestedSecondChild = nestedCmp.components().at(1).components().at(0);

    firstNestedRecord.set('user', 'updated_user1');
    expect(nestedFirstChild.get('name')).toBe('updated_user1');
    expect(nestedSecondChild.get('name')).toBe('nested_user2');
  });

  test('Nested collections are correctly serialized', () => {
    const serialized = parentCmp.toJSON();
    expect(serialized).toMatchSnapshot();
  });

  test('Nested collections respect startIndex and endIndex', () => {
    nestedCmpDef = {
      type: DataCollectionType,
      components: {
        type: DataCollectionItemType,
        components: {
          type: 'default',
          name: {
            type: DataVariableType,
            variableType: DataCollectionStateType.currentItem,
            collectionId: 'nested_collection',
            path: 'user',
          },
        },
      },
      dataResolver: {
        collectionId: 'nested_collection',
        startIndex: 0,
        endIndex: 1,
        dataSource: {
          type: DataVariableType,
          path: 'nested_data_source_id',
        },
      },
    };

    const updatedParentCmp = wrapper.components(getCmpDef(nestedCmpDef))[0] as unknown as ComponentDataCollection;
    const updatedNestedCmp = updatedParentCmp.getCollectionItemComponents().at(0) as ComponentDataCollection;
    expect(updatedNestedCmp.getItemsCount()).toBe(2);
  });

  test('Nested collection gets and watches value from the parent collection', () => {
    nestedCmpDef = {
      type: DataCollectionType,
      components: {
        type: DataCollectionItemType,
        components: {
          type: 'default',
          name: {
            type: DataVariableType,
            variableType: DataCollectionStateType.currentItem,
            collectionId: 'parent_collection',
            path: 'user',
          },
        },
      },
      dataResolver: {
        collectionId: 'nested_collection',
        startIndex: 0,
        endIndex: 1,
        dataSource: {
          type: DataVariableType,
          path: 'nested_data_source_id',
        },
      },
    };

    const updatedParentCmp = wrapper.components(getCmpDef(nestedCmpDef))[0] as unknown as ComponentDataCollection;
    const updatedNestedCmp = updatedParentCmp.getCollectionItemComponents().at(0) as ComponentDataCollection;
    const firstNestedChild = updatedNestedCmp.getCollectionItemComponents().at(0);

    expect(firstNestedChild.get('name')).toBe('user1');
    firstRecord.set('user', 'updated_user1');
    expect(firstNestedChild.get('name')).toBe('updated_user1');
  });

  test('Nested collection switches to using its own collection variable', () => {
    const firstChild = nestedCmp.components().at(0).components().at(0);

    firstChild.set('name', {
      // @ts-ignore
      type: DataVariableType,
      variableType: DataCollectionStateType.currentItem,
      path: 'user',
      collectionId: 'nested_collection',
    });

    expect(firstChild.get('name')).toBe('nested_user1');
  });

  describe('Nested Collection Component with Parent and Nested Data Sources', () => {
    beforeEach(() => {
      nestedCmpDef = {
        type: DataCollectionType,
        name: {
          type: DataVariableType,
          variableType: DataCollectionStateType.currentItem,
          collectionId: 'parent_collection',
          path: 'user',
        },
        components: {
          type: DataCollectionItemType,
          components: {
            type: 'default',
            name: {
              type: DataVariableType,
              variableType: DataCollectionStateType.currentItem,
              collectionId: 'nested_collection',
              path: 'user',
            },
          },
        },
        dataResolver: {
          collectionId: 'nested_collection',
          dataSource: {
            type: DataVariableType,
            path: 'nested_data_source_id',
          },
        },
      };

      parentCmp = wrapper.components(getCmpDef(nestedCmpDef))[0] as unknown as ComponentDataCollection;
      nestedCmp = parentCmp.getCollectionItemComponents().at(0) as ComponentDataCollection;
    });

    test('Removing a record from the parent data source updates the parent collection correctly', () => {
      expect(parentCmp.getItemsCount()).toBe(2);
      dataSource.removeRecord('user1');
      expect(parentCmp.getItemsCount()).toBe(1);
      expect(parentCmp.components().at(0).components().at(0).get('name')).toBe('user2');
      expect(nestedCmp.getItemsCount()).toBe(3);
      expect(nestedCmp.components().at(0).components().at(0).get('name')).toBe('nested_user1');
    });

    test('Adding a record to the parent data source updates the parent collection correctly', () => {
      expect(parentCmp.getItemsCount()).toBe(2);
      dataSource.addRecord({ id: 'user3', user: 'user3', age: '16' });
      expect(parentCmp.getItemsCount()).toBe(3);
      expect(parentCmp.components().at(2).components().at(0).get('name')).toBe('user3');
      expect(nestedCmp.getItemsCount()).toBe(3);
      expect(nestedCmp.components().at(0).components().at(0).get('name')).toBe('nested_user1');
    });

    test('Removing a record from the nested data source updates the nested collection correctly', () => {
      expect(nestedCmp.getItemsCount()).toBe(3);
      nestedDataSource.removeRecord('nested_user1');
      expect(nestedCmp.getItemsCount()).toBe(2);
      expect(nestedCmp.components().at(0).components().at(0).get('name')).toBe('nested_user2');
      expect(nestedCmp.components().at(1).components().at(0).get('name')).toBe('nested_user3');
    });

    test('Adding a record to the nested data source updates the nested collection correctly', () => {
      expect(nestedCmp.getItemsCount()).toBe(3);
      expect(nestedCmp.components().at(0).components().at(0).get('name')).toBe('nested_user1');
      expect(nestedCmp.components().at(1).components().at(0).get('name')).toBe('nested_user2');
      expect(nestedCmp.components().at(2).components().at(0).get('name')).toBe('nested_user3');

      nestedDataSource.addRecord({ id: 'user4', user: 'nested_user4', age: '18' });
      expect(nestedCmp.getItemsCount()).toBe(4);
      expect(nestedCmp.components().at(3).components().at(0).get('name')).toBe('nested_user4');
      expect(nestedCmp.components().at(0).components().at(0).get('name')).toBe('nested_user1');
      expect(nestedCmp.components().at(1).components().at(0).get('name')).toBe('nested_user2');
      expect(nestedCmp.components().at(2).components().at(0).get('name')).toBe('nested_user3');
    });
  });
});
