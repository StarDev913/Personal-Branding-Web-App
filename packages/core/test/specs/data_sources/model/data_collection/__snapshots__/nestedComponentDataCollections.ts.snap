// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection component Nested collections are correctly serialized 1`] = `
{
  "components": [
    {
      "components": [
        {
          "components": [
            {
              "components": [
                {
                  "name": {
                    "collectionId": "nested_collection",
                    "path": "user",
                    "type": "data-variable",
                    "variableType": "currentItem",
                  },
                  "type": "default",
                },
              ],
              "type": "data-collection-item",
            },
          ],
          "dataResolver": {
            "collectionId": "nested_collection",
            "dataSource": {
              "path": "nested_data_source_id",
              "type": "data-variable",
            },
          },
          "type": "data-collection",
        },
      ],
      "type": "data-collection-item",
    },
  ],
  "dataResolver": {
    "collectionId": "parent_collection",
    "dataSource": {
      "path": "my_data_source_id",
      "type": "data-variable",
    },
  },
  "type": "data-collection",
}
`;
